AWS AL2023 support

nbusseneau opened the pull request:
This PR aims to introduce transparent support for installing Cilium on clusters using the AL2023 AWS node image family.

Please see per commit for full details :)

nbusseneau committed:
cli: move AWS chaining mode setup to AWS-specific code

We move the logic for setting up chaining mode on AWS from the high
level entrypoint to the dedicated AWS-specific code. This will help keep
the entrypoint clean as we intend to enrich the high level logic for AWS
in subsequent commits.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

nbusseneau committed:
cli: add AWS node image family detection mechanism

On AL2023 images, interfaces use the `ens` naming scheme instead of the
older `eth` one. Separatley, `eksctl` makes AL2023 the default image for
new Kubernetes clusters using Kubernetes 1.30 and above, with AL2 still
being the default on 1.29 and below.

Since we were previously assuming the `eth` naming scheme, trying to
install Cilium on a newly created `eksctl` cluster with K8s 1.30 or
above would not work, and one would have to choose one of the following
manual workaround:

- Stick to using AL2 images.
- Reconfigure Cilium to work with `ens` interfaces.
- Manually change the interface naming scheme.

This commit introduces an auto-detection mechanism for the node image
family currently available on the cluster, allowing us to properly
configure Cilium to use either `eth` or `ens` interfaces depending on
if AL2023 is in use or not.

We stick to `eth` naming for anything non-AL2023 for now, for backwards
compatibility, but this new capability may prove useful in the future
to differentiate on even more image families (e.g. Bottlerocket or
Ubuntu).

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

auriaave commented on cilium-cli/install/helm.go:
```diff
@@ -64,8 +64,12 @@ func (k *K8sInstaller) getHelmValues() (map[string]interface{}, error) {
+			// AL2023 uses ens interfaces, but we default to eth interfaces for everything else for backwards compatibility,
+			// since the CLI was assuming eth interfaces before support for AL2023 was introduced
 			helmMapOpts["egressMasqueradeInterfaces"] = "eth0"
+			if k.params.AWS.AwsNodeImageFamily == AwsNodeImageFamilyAmazonLinux2023 {
```
Do we know the interface naming pattern on all the image families listed (except custom)? Are all of them following eth0, except AL2023? Is it possible to retrieve the interface name directly from the host rather than infer it by the image family used?

The reason I ask is in the case where custom images based on AL2023 are used, or in the future other images follow the same convention.

nbusseneau replied:
Neither the AWS CLI nor eksctl seem to expose this information, so it indeed would have to be retrieved from the host directly. There's the `ExecInPod` facility that could probably be used for that -- we can execute `ip address` commands and parse the output, just need a pod to leverage -- that would probably be the annoying part, since we are executing this **before** Cilium is installed.

auriaave replied:
we can leave this as a future enhancement. Thanks for checking.

auriaave replied:
This suggestion can be closed. Not pursuing it for now.

auriaave commented on cilium-cli/install/install.go:
```diff
@@ -224,18 +223,14 @@ func (k *K8sInstaller) preinstall(ctx context.Context) error {
+		}
+
+		// setup chaining mode
+		if err := k.awsSetupChainingMode(ctx, helmValues); err != nil {
```
This part has only been moved to the aws.go file, correct? No changes impacting the above retrieval of the node image familly.

nbusseneau replied:
Correct -- I have split it in a separate commit for easier review of the move change.

auriaave reviewed (commented):
Possibly add unit tests for aws.go - otherwise, minor comments. Thanks.

nbusseneau committed:
docs: add small note regarding AL2023 in Helm install guide

AL2023 uses `ens5` instead of `eth0` as AL2 did, so we add a small
snippet about this in the docs. Eventually we might want to change the
docs to have `ens5` by default, if we assume most users would be using
AL2023 images for their clusters.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

gandro commented on cilium-cli/install/helm.go:
```diff
@@ -64,8 +64,12 @@ func (k *K8sInstaller) getHelmValues() (map[string]interface{}, error) {
+			// since the CLI was assuming eth interfaces before support for AL2023 was introduced
 			helmMapOpts["egressMasqueradeInterfaces"] = "eth0"
+			if k.params.AWS.AwsNodeImageFamily == AwsNodeImageFamilyAmazonLinux2023 {
+				helmMapOpts["egressMasqueradeInterfaces"] = "ens5"
```
How certain are we that `ens5` is the correct interface on all nodes? Should we use a prefix match such as `ens+` here? I believe even the previous `eth0` probably should have been `eth+` to be honest, i.e. pod-to-pod traffic on secondary interfaces probably should also be subject to masquerading. 

Unfortunately our docs are a bit conflicting here. The ENI docs do recommend `eth+`, where as the Helm getting started guide just sets `eth0`:

https://docs.cilium.io/en/v1.16/network/concepts/routing/#id6
https://docs.cilium.io/en/latest/installation/k8s-install-helm/#k8s-install-helm

gandro reviewed (commented):
Some inline question on the interface name

viktor-kurchenko commented on cilium-cli/install/aws.go:
```diff
@@ -3,9 +3,99 @@
+		return err
+	}
+
+	clusterInfo := awsClusterInfo{}
```
I think `eksctl get nodegroup` command will return output as JSON array, isn't it?
If yes then `json.Unmarshal` into a struct will fail.

viktor-kurchenko committed:
cli: AWS node image family detection refactoring

Use K8S API instead of `eksctl` tool to list cluster nodes and
detect OS image.
Current implementation does not support clusters with different
node images (e.g.: `Amazon Linux 2` and `Amazon Linux 2023`).
However different versions of same node image (e.g.:
`Amazon Linux 2023.1.20230823` and `Amazon Linux 2023.6.20241031`)
is allowed.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

viktor-kurchenko committed:
cli: egress masquerade selector improved

Egress masquerade interfaces selector logic improved to match interfaces
by prefix.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

auriaave commented on .github/actions/aws-cni/k8s-versions.yaml:
```diff
@@ -1,15 +1,15 @@
     default: true
     kpr: true
-  - version: "1.29"
+  - version: "1.30"
```
The current file supports up to EKS 1.31 - be careful not to regress 🙏

viktor-kurchenko replied:
Yeah, it's a `temp commit DO NOT MERGE` [commit](https://github.com/cilium/cilium/pull/35529/commits/d4d21e43e2cd18688f14ce56931f66ae88595ef1).
I'll remove it before merge.

viktor-kurchenko replied:
Removed.

auriaave commented on .github/actions/eks/k8s-versions.yaml:
```diff
@@ -1,17 +1,17 @@
     ipsec: true
     kpr: true
-  - version: "1.29"
+  - version: "1.30"
```
Same about https://github.com/cilium/cilium/blob/main/.github/actions/eks/k8s-versions.yaml

viktor-kurchenko replied:
Yeah, it's a `temp commit DO NOT MERGE` [commit](https://github.com/cilium/cilium/pull/35529/commits/d4d21e43e2cd18688f14ce56931f66ae88595ef1).
I'll remove it before merge.

viktor-kurchenko replied:
Removed.

auriaave commented on cilium-cli/install/aws.go:
```diff
@@ -3,9 +3,111 @@
+	// because currently mixed nodes setup is not supported
+	for i := 1; i < len(nodes); i++ {
+		if ami != amiFn(nodes[i].Status.NodeInfo.OSImage) {
+			return "", errors.New("cluster has nodes with different OS images")
```
Can we possibly add the 2 values that don't match? The "ami" one and the one from nodes[i]? It will contribute to add more info while troubleshooting the problem.

viktor-kurchenko replied:
Done, thanks!

auriaave reviewed (approved):
Added a suggestion to include the mismatching OS images found in the error message. Not a blocking issue. Thanks.

learnitall reviewed (approved):
Docs look good to go! :rocket:

derailed reviewed (approved):
@nbusseneau Nice work!
